File: examples/02-intro-cpp/example10.cpp
// =================================================================
//
// File: example10.cpp
// Author: Pedro Perez
// Description: This file implements the code that blurs a given
//				image. Uses OpenCV, to compile:
//				g++ example10.cpp `pkg-config --cflags --libs opencv4`
//
//				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/highgui/highgui_c.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/imgcodecs/imgcodecs.hpp>
#include "utils.h"

#define BLUR_WINDOW 15

class BlurImage {
private:
	cv::Mat &src, &dest;

	void blurPixel(int ren, int col) const {
		int side_pixels, cells;
		int tmp_ren, tmp_col;
		float r, g, b;

		side_pixels = (BLUR_WINDOW - 1) / 2;
		cells = (BLUR_WINDOW * BLUR_WINDOW);
		r = 0; g = 0; b = 0;
		for (int i = -side_pixels; i <= side_pixels; i++) {
			for (int j = -side_pixels; j <= side_pixels; j++) {
				tmp_ren = MIN( MAX(ren + i, 0), src.rows - 1);
				tmp_col = MIN( MAX(col + j, 0), src.cols - 1);

				r += (float) src.at<cv::Vec3b>(tmp_ren, tmp_col)[RED];
				g += (float) src.at<cv::Vec3b>(tmp_ren, tmp_col)[GREEN];
				b += (float) src.at<cv::Vec3b>(tmp_ren, tmp_col)[BLUE];
			}
		}

		dest.at<cv::Vec3b>(ren, col)[RED] =  (unsigned char) (r / cells);
		dest.at<cv::Vec3b>(ren, col)[GREEN] = (unsigned char) (g / cells);
		dest.at<cv::Vec3b>(ren, col)[BLUE] = (unsigned char) (b / cells);
	}

public:
	BlurImage(cv::Mat &s, cv::Mat &d) : src(s), dest(d) {}

	void doTask() {
		for(int i = 0; i < src.rows; i++) {
			for(int j = 0; j < src.cols; j++) {
				blurPixel(i, j);
			}
		}
	}
};

int main(int argc, char* argv[]) {
	int i;
	double acum;

	if (argc != 2) {
		printf("usage: %s source_file\n", argv[0]);
		return -1;
	}

	cv::Mat src = cv::imread(argv[1], cv::IMREAD_COLOR);
	cv::Mat dest = cv::Mat(src.rows, src.cols, CV_8UC3);
	if (!src.data) {
		printf("Could not load image file: %s\n", argv[1]);
		return -1;
	}

	acum = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		BlurImage obj(src, dest);
		obj.doTask();

		acum += stop_timer();
	}

	printf("avg time = %.5lf ms\n", (acum / N));

	/*
	cv::namedWindow("Original", cv::WINDOW_AUTOSIZE);
	cv::imshow("Original", src);

	cv::namedWindow("Blur", cv::WINDOW_AUTOSIZE);
	cv::imshow("Blur", dest);

	cv::waitKey(0);
	*/

	cv::imwrite("blur.png", dest);

	return 0;
}

File: examples/04-fork-join/Example10/Example10.java
// =================================================================
//
// File: Example10.java
// Author: Pedro Perez
// Description: This file implements the code that blurs a given
//				image using Java's Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;
import java.io.IOException;
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class Example10 extends RecursiveAction {
	private static final int MIN = 15_000;
	private static final int BLUR_WINDOW = 15;
	private int src[], dest[], width, height, start, end;

	public Example10(int src[], int dest[], int width, int height, int start, int end) {
		this.src = src;
		this.dest = dest;
		this.width = width;
		this.height = height;
		this.start = start;
		this.end = end;
	}

	private void blurPixel(int ren, int col) {
		int side_pixels, i, j, cells;
		int tmp_ren, tmp_col, pixel, dpixel;
		float r, g, b;

		side_pixels = (BLUR_WINDOW - 1) / 2;
		cells = (BLUR_WINDOW * BLUR_WINDOW);
		r = 0; g = 0; b = 0;
		for (i = -side_pixels; i <= side_pixels; i++) {
			for (j = -side_pixels; j <= side_pixels; j++) {
				tmp_ren = Math.min( Math.max(ren + i, 0), height - 1 );
				tmp_col = Math.min( Math.max(col + j, 0), width - 1);
				pixel = src[(tmp_ren * width) + tmp_col];

				r += (float) ((pixel & 0x00ff0000) >> 16);
				g += (float) ((pixel & 0x0000ff00) >> 8);
				b += (float) ((pixel & 0x000000ff) >> 0);
			}
		}

		dpixel = (0xff000000)
				| (((int) (r / cells)) << 16)
				| (((int) (g / cells)) << 8)
				| (((int) (b / cells)) << 0);
		dest[(ren * width) + col] = dpixel;
	}

	protected void computeDirectly() {
		int index, ren, col;

		for (index = start; index < end; index++) {
			ren = index / width;
			col = index % width;
			blurPixel(ren, col);
		}
	}

	@Override
	protected void compute() {
		if ( (end - start) <= MIN ) {
			computeDirectly();
		} else {
			int mid = start + ((end - start) / 2);
			invokeAll(new Example10(src, dest, width, height, start, mid),
					  new Example10(src, dest, width, height, mid, end));
		}
	}

	public static void main(String args[]) throws Exception {
		long startTime, stopTime;
		double ms;
		int src[], dest[], w, h;
		ForkJoinPool pool;

		if (args.length != 1) {
			System.out.println("usage: java Example10 image_file");
			System.exit(-1);
		}

		final String fileName = args[0];
		File srcFile = new File(fileName);
        final BufferedImage source = ImageIO.read(srcFile);

		w = source.getWidth();
		h = source.getHeight();
		src = source.getRGB(0, 0, w, h, null, 0, w);
		dest = new int[src.length];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			pool.invoke(new Example10(src, dest, w, h, 0, (w * h)));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
		}

		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
		final BufferedImage destination = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
		destination.setRGB(0, 0, w, h, dest, 0, w);

		/*
		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Original - " + fileName, source);
            }
        });

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Blur - " + fileName, destination);
            }
        });
		*/

		try {
			ImageIO.write(destination, "png", new File("blur.png"));
			System.out.println("Image was written succesfully.");
		} catch (IOException ioe) {
			System.out.println("Exception occured :" + ioe.getMessage());
		}
	}
}
